x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTrackerParameterizedTests.java

```diff
@@ -0,0 +1,1115 @@
+            nodeMemorySizes.size(),
+            nodeProcessorSizes.size()
+        );
+
+        int smallestSufficientNodeIndex = nodeMemorySizes.size();
+
+        for (int i = 0; i < nodeMemorySizes.size(); i++) {
+            if (nodeMemorySizes.get(i).getBytes() >= minMemoryPerNode.getBytes() && nodeProcessorSizes.get(i) >= minProcessorsPerNode) {
+                smallestSufficientNodeIndex = i;
+                break;
+            }
+        }
+
+        double numProcessorsPerNode = nodeProcessorSizes.get(smallestSufficientNodeIndex);
+        ByteSizeValue memoryPerNode = nodeMemorySizes.get(smallestSufficientNodeIndex);
+
+        int assignedProcessors = 0;
+        ByteSizeValue assignedMemory = ByteSizeValue.ZERO;
+        DiscoveryNodes.Builder dnBuilder = DiscoveryNodes.builder();
+        int nodeCount = 0;
+
+        while (assignedProcessors < totalProcessors || assignedMemory.getBytes() < totalMemory.getBytes()) {
+            dnBuilder.add(buildDiscoveryNode(numProcessorsPerNode, String.valueOf(memoryPerNode.getBytes()), nodeCount));
+            assignedProcessors += (int) numProcessorsPerNode;
+            assignedMemory = ByteSizeValue.add(memoryPerNode, assignedMemory);
+            nodeCount += 1;
+        }
+
+        return dnBuilder.build();
+    }
+
+    private static DiscoveryNode buildDiscoveryNode(double numProcessorsPerNode, String memoryPerNode, int nodeNumber) {
+        Map<String, String> attributes = Map.of(
+            MACHINE_MEMORY_NODE_ATTR,
+            memoryPerNode,
+            MachineLearning.ALLOCATED_PROCESSORS_NODE_ATTR,
+            String.valueOf(numProcessorsPerNode),
+            MAX_JVM_SIZE_NODE_ATTR,
+            Long.toString(Runtime.getRuntime().maxMemory())
+        );
+
+        return DiscoveryNodeUtils.builder(NODE_NAME_PREFIX + nodeNumber)
+            .attributes(attributes)
+            .roles(Set.of(DiscoveryNodeRole.ML_ROLE))
+            .build();
+    }
+
+    private static ClusterSettings createClusterSettings() {
+        return new ClusterSettings(Settings.EMPTY, Set.of(MachineLearning.ALLOCATED_PROCESSORS_SCALE));
+    }
+
+    private static MlMemoryTracker createMlMemoryTracker() {
+        return mock(MlMemoryTracker.class);
+    }
+
+    private static Settings createSettings() {
+        return Settings.builder()
+            .put(MAX_MACHINE_MEMORY_PERCENT.getKey(), MAX_MACHINE_MEMORY_PERCENT.get(Settings.EMPTY))
+            .put(USE_AUTO_MACHINE_MEMORY_PERCENT.getKey(), USE_AUTO_MACHINE_MEMORY_PERCENT.get(Settings.EMPTY))
+            .build();
+    }
+
+    private static StartTrainedModelDeploymentAction.TaskParams createTaskParams(int numAllocations, int seed) {
+
+        String modelId = "modelId" + seed;
+        String deploymentId = "deploymentId" + seed;
+        long modelBytes = MODEL_BYTES;
+        int threadsPerAllocation = 1; // TODO expand to multiple threads per allocation
```

Comment:
Would be nice to include at least one test with `threadPerAllocation > 1`

Reply:
I duplicated all of the test cases to include a test with threadPerAllocation=2